Userland/Utilities/headless-browser.cpp

```diff
@@ -264,8 +277,216 @@ class HeadlessImageDecoderClient : public Web::ImageDecoding::Decoder {
+            auto underlying_socket = TRY(Core::Stream::TCPSocket::connect(url.host(), url.port().value_or(80)));
+            TRY(underlying_socket->set_blocking(false));
+            auto socket = TRY(Core::Stream::BufferedSocket<Core::Stream::TCPSocket>::create(move(underlying_socket)));
+
+            HTTP::HttpRequest request;
+            if (method.equals_ignoring_case("head"sv))
+                request.set_method(HTTP::HttpRequest::HEAD);
+            if (method.equals_ignoring_case("get"sv))
+                request.set_method(HTTP::HttpRequest::GET);
+            if (method.equals_ignoring_case("post"sv))
+                request.set_method(HTTP::HttpRequest::Invalid);
+            request.set_url(move(url));
+            request.set_headers(request_headers);
+            request.set_body(TRY(ByteBuffer::copy(request_body)));
+
+            return adopt_ref(*new HTTPHeadlessRequest(move(request), move(socket), move(stream_backing_buffer)));
+        }
+
+        virtual ~HTTPHeadlessRequest() override
+        {
+        }
+
+        virtual void set_should_buffer_all_input(bool) override
+        {
+        }
+
+    private:
+        HTTPHeadlessRequest(HTTP::HttpRequest&& request, NonnullOwnPtr<Core::Stream::BufferedSocketBase> socket, ByteBuffer&& stream_backing_buffer)
+            : m_stream_backing_buffer(move(stream_backing_buffer))
+            , m_output_stream(Core::Stream::MemoryStream::construct(m_stream_backing_buffer.bytes()).release_value_but_fixme_should_propagate_errors())
+            , m_socket(move(socket))
+            , m_job(HTTP::Job::construct(move(request), *m_output_stream))
+        {
+            m_job->on_headers_received = [weak_this = make_weak_ptr()](auto& response_headers, auto response_code) mutable {
+                dbgln("Headers received");
+                if (!weak_this)
+                    return;
+                weak_this->m_response_code = response_code;
+                for (auto& header : response_headers) {
+                    weak_this->m_response_headers.set(header.key, header.value);
+                }
+            };
+            m_job->on_finish = [weak_this = make_weak_ptr()](bool success) mutable {
+                dbgln("Request finished");
+                if (!weak_this)
+                    return;
+                Core::deferred_invoke([weak_this, success]() mutable {
+                    dbgln("Request resolved!!!");
+                    if (!weak_this)
+                        return;
+                    ReadonlyBytes response_bytes { weak_this->m_output_stream->bytes().data(), weak_this->m_output_stream->offset() };
+                    auto response_buffer = ByteBuffer::copy(response_bytes).release_value_but_fixme_should_propagate_errors();
+                    weak_this->on_buffered_request_finish(success, weak_this->m_output_stream->offset(), weak_this->m_response_headers, weak_this->m_response_code, response_buffer);
```

Comment:
Is there no data race here with weak_this? I think you want to strong_ref for the duration of the callback.